### PR TITLE
Surface implicit privileges in get-role API

### PR DESCRIPTION
**Pull request by ebarlas**
The implicit-privileges SPI lets extensions inject index privileges derived from a role's application privileges at role-build time, but the get-role API still returns only the explicitly stored privileges, leaving Kibana and other admin tools no way to show users what they will actually be authorized to do.

Add an opt-in `include_implicit` query parameter that decorates each returned `RoleDescriptor` with the `IndicesPrivileges` its registered `ImplicitPrivilegesProviders` would contribute. The default is `false` so existing clients see no behavior change.

Each implicit entry is wrapped in a new `ImplicitlyGranted` marker subclass that emits `implicitly_granted: true` at xcontent time only. The PUT parser rejects `implicitly_granted` as an unknown field so a GET-then-PUT round-trip cannot accidentally promote an implicit grant into a stored one.

When `xpack.security.dls_fls.enabled` is false, implicit entries that carry DLS or FLS are dropped from the response to mirror the runtime suppression in `addImplicitPrivileges`, keeping the response in lock-step with what authorization enforces.

A new `x-pack/qa/security-implicit-privileges-extension` module wires `HelicarrierImplicitPrivilegesProvider` through the `SecurityExtension` SPI so the `include_implicit` parameter and the `implicitly_granted` response field can be asserted end-to-end against a real cluster.

### CODE REVIEW

File: rest-api-spec/src/main/resources/rest-api-spec/api/security.get_role.json [hunk 1]
```diff
@@ -32,6 +32,13 @@
           ]
         }
       ]
+    },
+    "params": {
+      "include_implicit": {
+        "type": "boolean",
+        "default": false,
+        "description": "If true, include privileges that are implicitly granted by registered ImplicitPrivilegesProviders alongside the explicitly configured privileges."
```
Comment (n1v0lg) [inline rest-api-spec/src/main/resources/rest-api-spec/api/security.get_role.json:40]: 👍 lets also add it in the [spec](https://github.com/elastic/elasticsearch-specification/blob/main/specification/security/get_role/SecurityGetRoleRequest.ts#L35)

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/CompositeRolesStore.java [hunk 2]
```diff
@@ -722,6 +711,91 @@ private static void addImplicitPrivileges(
...
+        Collection<RoleDescriptor> roleDescriptors,
+        Map<Tuple<String, String>, ApplicationPrivilegeDescriptor> appPrivLookup
+    ) {
+        List<RoleDescriptor> result = new ArrayList<>();
+        for (RoleDescriptor rd : roleDescriptors) {
+            List<ApplicationPrivilegeDescriptor> roleAppPrivs = getApplicationPrivilegeDescriptors(rd, appPrivLookup);
+            List<IndicesPrivileges> implicitPrivileges = implicitPrivilegesProviders.stream()
+                .flatMap(p -> p.getImplicitIndicesPrivileges(rd, roleAppPrivs).stream())
+                .<IndicesPrivileges>map(IndicesPrivileges.ImplicitlyGranted::new)
+                .filter(ip -> dlsFlsEnabled || !ip.isUsingDocumentOrFieldLevelSecurity())
```
Comment (n1v0lg) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/CompositeRolesStore.java:772]: The second part of this condition (`!ip.isUsingDocumentOrFieldLevelSecurity()`) tripped me up a bit -- could you add a comment along the lines of:

`// always include implicit permissions that do not rely on DLS/FLS; omit ones that do if DLS/FLS is disabled`

What tripped me up is that we can in principle have implicit permissions that don't rely on DLS/FLS (I had assumed -- without good reason -- that they always do).

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/CompositeRolesStore.java [hunk 3]
```diff
@@ -722,6 +711,91 @@ private static void addImplicitPrivileges(
...
+        }
+        Set<String> applicationNames = roleDescriptors.stream()
+            .flatMap(rd -> Stream.of(rd.getApplicationPrivileges()))
+            .map(RoleDescriptor.ApplicationResourcePrivileges::getApplication)
+            .collect(Collectors.toSet());
+        Set<String> privilegeNames = roleDescriptors.stream()
+            .flatMap(rd -> Stream.of(rd.getApplicationPrivileges()))
+            .flatMap(arp -> Stream.of(arp.getPrivileges()))
+            .collect(Collectors.toSet());
+        if (applicationNames.isEmpty() || privilegeNames.isEmpty()) {
```
Comment (n1v0lg) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/CompositeRolesStore.java:750]: I don't think this is introduced in this PR but I think there's an edge case worth considering for implicit privileges more generally:

An admin role can have

```
{
  "applications": [
    {
      "application": "*",
      "privileges": ["*"],
      "resources": ["*"]
    }
  ]
}
```

application privileges. I'd expect that to translate to implicit privileges. However, that's not the case, see:

```java
public void testWildcardApplicationPrivilegeDoesNotResolveStoredPrivilegesForProvider() throws Exception {
    putShieldAgentApplicationPrivilege();
    final String roleName = "wildcard-app-role";
    final Request put = new Request("PUT", "/_security/role/" + roleName);
    put.setJsonEntity("""
        {
          "cluster": ["monitor"],
          "indices": [
            { "names": ["logs-*"], "privileges": ["read"] }
          ],
          "applications": [
            { "application": "*", "privileges": ["*"], "resources": ["*"] }
          ]
        }
        """);
    assertOK(client().performRequest(put));
    // The role holds *, * application privileges which effectively grant all application
    // actions at authorization time via pattern matching. However, action patterns like "*"
    // are not valid stored-privilege names and NativePrivilegeStore short-circuits to an
    // empty result set. The provider therefore never sees the shield/agent descriptor and
    // does not contribute implicit index privileges.
    final RoleResponse response = getRole(roleName, true);
    assertThat(response.indices(), hasSize(1));
    response.assertNoImplicitlyGrantedEntries();
    assertThat(response.indices().get(0).names(), equalTo(List.of("logs-*")));
}
```

Something to consider for a follow up. 

Note that a user that has `{ "application": "*", "privileges": ["*"], "resources": ["*"] }` is very likely to also have `*` index `read` or higher so this is certainly more of an edge case.

**Review by n1v0lg — COMMENTED** [review]
Production code looks good 👍 nice work. Just a few comments. I'll review test code tomorrow.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/CompositeRolesStore.java (hunk 3 above)
Comment (ebarlas) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/CompositeRolesStore.java:750]: Great point! A bit more work is needed to properly support wildcards in application names and privilege names generally. I'll work on that and provide and update later.

Comment (n1v0lg) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/CompositeRolesStore.java:750]: 👍 non-blocking concern for this PR

Comment (ebarlas) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/CompositeRolesStore.java:750]: @n1v0lg , I've given this some thought and I have a few things to report regarding wildcard support.

#### Background
Notes about `RoleDescriptor` application privileges contents:
- Application names can optionally have a single trailing wildcard (e.g. `myapp*`, `*`, etc)
- Privilege names can either be (1) stored application privilege names (e.g `read`, `write`) or (2) action-patterns (e.g. `ui:read`, `ui:*`)

#### Issue in `CompositeRolesStore`
There is an omission in the production code: `CompositeRolesStore` doesn't consider wildcards that appear in application names in `RoleDescriptor` application privileges. For example, a `RoleDescriptor` with application `kibana-*` and privilege `read` won't resolve to an `ApplicationPrivilegeDescriptor`.

The fix is straightforward but it does invalidate the simple lookup table keyed by (app-name, priv-name) in use now in `CompositeRolesStore#getApplicationPrivilegeDescriptors`. Instead, application names _with a wildcard_ need to be checked against each `ApplicationPrivilegeDescriptor`.

#### Note About SPI Implementation Scope
Action patterns create an additional burden for `ImplicitPrivilegesProvider` implementations that are concerned with specific actions. Those impls need to scan (1) `ApplicationPrivilegeDescriptor` instances looking for action matches and (2) `RoleDescriptor` application privileges looking for action patterns.

#### What's Next?
A straightforward code change in `CompositeRolesStore` adds wildcard awareness. I'll do that following this PR.

Comment (n1v0lg) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/CompositeRolesStore.java:750]: Thanks for investigating -- that summary tracks for me. ++ on a follow up for this

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/store/CompositeRolesStoreTests.java [hunk 4]
```diff
@@ -1187,6 +1187,310 @@ public void testImplicitPrivilegesOverSameIndicesAreCollatedIntoOneGroup() {
...
+        final Collection<RoleDescriptor> result = invokeAddImplicitPrivileges(store, List.of(rd));
+
+        assertThat(result, hasSize(1));
+        // Role had no explicit indices privileges; the result must be exactly one entry, and it
+        // must be the ImplicitlyGranted marker subclass — that is what gates the
+        // "implicitly_granted":true field at xcontent rendering time.
+        final IndicesPrivileges[] indices = result.iterator().next().getIndicesPrivileges();
+        assertThat(indices, arrayWithSize(1));
+        assertThat(indices[0], instanceOf(IndicesPrivileges.ImplicitlyGranted.class));
+        assertThat(indices[0].getIndices(), arrayContaining("helicarrier-*"));
```
Comment (n1v0lg) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/store/CompositeRolesStoreTests.java:1209]: Would be nice to also cover a case where a role has the same exact explicit privilege as implicit privilege.

File: x-pack/qa/security-implicit-privileges-extension/src/javaRestTest/java/org/elasticsearch/example/implicit/ImplicitPrivilegesIT.java [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+        assertThat(implicit.query(), containsString("clearance"));
+    }
+
+    public void testIncludeImplicitTrueWithoutQualifyingApplicationPrivilegeLeavesResponseUnchanged() throws Exception {
+        // No application privilege referenced, so the provider returns nothing. The response
+        // shape must be identical to include_implicit=false: no extra entries, no marker.
+        final String roleName = "ops-engineer";
+        putRoleWithoutApplicationPrivilege(roleName);
+
+        final RoleResponse response = getRole(roleName, true);
```
Comment (n1v0lg) [inline x-pack/qa/security-implicit-privileges-extension/src/javaRestTest/java/org/elasticsearch/example/implicit/ImplicitPrivilegesIT.java:106]: Would be nice to either randomly (or in a separate block/test) also cover the case where user has an application privilege but not one that grants an implicit privilege.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/store/CompositeRolesStoreTests.java [hunk 6]
```diff
@@ -1187,6 +1187,310 @@ public void testImplicitPrivilegesOverSameIndicesAreCollatedIntoOneGroup() {
...
+            mock(ServiceAccountService.class),
+            projectResolver,
+            buildBitsetCache(),
+            TestRestrictedIndices.RESTRICTED_INDICES,
+            mockRoleBuildingExecutor,
+            rds -> {},
+            providers
+        );
+    }
+
```
Comment (n1v0lg) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/store/CompositeRolesStoreTests.java:1543]: Might also cover the case where `NativePrivilegeStore#getPrivileges()` throws.

**Review by n1v0lg — APPROVED** [review]
LGTM -- a few non-blocking test suggestions.

File: rest-api-spec/src/main/resources/rest-api-spec/api/security.get_role.json (hunk 1 above)
Comment (ebarlas) [reply rest-api-spec/src/main/resources/rest-api-spec/api/security.get_role.json:40]: Spec PR: https://github.com/elastic/elasticsearch-specification/pull/6255
